### PR TITLE
no-codegen compile flag and global generics reflection

### PR DESCRIPTION
**Pull request by csyonghe**
1. Add `SLANG_COMPILE_FLAG_NO_CODEGEN` (`-no-codegen`) compiler flag to skip code generation stage, so that a shader that uses global generic type parameters can be parsed, checked and introspected without knowing the final specialization.

2. Add reflection API to query for global generic type parameters, global parameters of generic type, and the generic type parameter index related to a global generic parameter.

3. Add a reflection test case for global generic type parameters.

### CODE REVIEW

**Comment by tangent-vector** [comment]
Looks like a bunch of tests are failing. You'll probably want to run the tests locally and diagnose the failures.

In the case of the reflection tests, you probably want to confirm that your new code prints nothing new if there are no type parameters (rather than printing an empty list).

**Comment by csyonghe** [comment]
Ah, yeah, I am printing an empty type params array for those test cases, thus causing the failures.

**Comment by csyonghe** [comment]
Actually it's because I changed the '\n' output...
Fixing it now.

**Comment by tfoleyNV** [comment]
I think you are also getting failures on compute tests, so there seems to be more that needs to be fixed.

**Comment by csyonghe** [comment]
Should be fixed now. I am a bad programmer writing opposite if conditions.

File: source/slang/slang.cpp [hunk 1]
```diff
@@ -293,9 +293,12 @@ int CompileRequest::executeActionsInner()
         if (mSink.GetErrorCount() != 0)
             return 1;
 
-        // Generate initial IR for all the translation
-        // units, if we are in a mode where IR is called for.
-        generateIR();
+        if ((compileFlags & SLANG_COMPILE_FLAG_NO_CODEGEN) == 0)
+        {
```
Comment (tfoleyNV) [inline source/slang/slang.cpp]: This is fine for now, but I think that eventually we need/want to generate the “initial” IR even if target code generation is disabled. That is, the output of the first compilation phase should be IR + reflection, which feeds into the second phase.

**Pull request merged**